File: java/perfect-numbers/src/main/java/NaturalNumber.java
import java.util.stream.IntStream;

class NaturalNumber {
    private final int number;

    NaturalNumber(int number) {
        if (number <= 0) {
            throw new IllegalArgumentException("You must supply a natural number (positive integer)");
        }
        this.number = number;
    }

    Classification getClassification() {
        int aliquot = getAliquotSum();
        if (aliquot < number) {
            return Classification.DEFICIENT;
        } else if (aliquot > number) {
            return Classification.ABUNDANT;
        } else {
            return Classification.PERFECT;
        }
    }

    private int getAliquotSum() {
        return IntStream.rangeClosed(1, number / 2).filter(i -> number % i == 0).sum();
    }
}

File: kotlin/perfect-numbers/src/main/kotlin/NaturalNumber.kt

enum class Classification {
    DEFICIENT, PERFECT, ABUNDANT
}

fun classify(naturalNumber: Int): Classification {
    if (naturalNumber < 1) {
        throw IllegalArgumentException("Natural numbers are >= 1.")
    }
    val alSum = aliquotSum(naturalNumber)
    return when {
        alSum < naturalNumber -> Classification.DEFICIENT
        alSum > naturalNumber -> Classification.ABUNDANT
        else -> Classification.PERFECT
    }
}

fun aliquotSum(num: Int): Int =  (1 until num).filter { num % it == 0 }.sum()
